### PR TITLE
Saving Princess: implement new game

### PR DESCRIPTION
**Pull request by LeonarthCG**
## What is this fixing or adding?

Adds support for Saving Princess to Archipelago.

## How was this tested?

The world was tested with an async that had 3 players run the game, together with another 6 games. Only two major issues were found during this, which have since been fixed.

Over the past month and a half I have also played seeds by myself, testing every option by running the game on its own, with other slots of the game and also with other supported and unsupported games.

## If this makes graphical changes, please attach screenshots.
image

### CODE REVIEW

File: worlds/saving_princess/__init__.py
Comment (ScipioWright) [inline worlds/saving_princess/__init__.py:106]:
```suggestion
        if self.random.randint(0, 99) < self.options.trap_chance:
            filler_list = list(Items.item_dict_traps.keys())
        # and return one of the names at random
        return self.random.choice(filler_list)
```

Would recommend using the world's random object instead of the multiworld random object. Makes it more deterministic, basically.

File: worlds/saving_princess/docs/en_Saving Princess.md
Comment (ScipioWright) [inline worlds/saving_princess/docs/en_Saving Princess.md:3]:
```suggestion
## Where is the options page?
```

File: worlds/saving_princess/docs/setup_en.md
Comment (ScipioWright) [inline worlds/saving_princess/docs/setup_en.md:48]:
```suggestion
You can customize your options by visiting the [Saving Princess Player Options Page](/games/Saving Princess/player-options)
```

Settings was changed to Options a while back

File: worlds/saving_princess/Regions.py
Comment (ScipioWright) [inline worlds/saving_princess/Regions.py:20]: Is there a specific reason this is a `Dict[str, RegionData]` rather than a `Dict[str, List[str]]`? The surrounding class feels unnecessary, but maybe I'm just missing something here.

File: worlds/saving_princess/Regions.py
Comment (ScipioWright) [inline worlds/saving_princess/Regions.py:105]:
```suggestion
def create_regions(multiworld: MultiWorld, player: int, is_pool_expanded: bool):
    for region_name, region_data in region_dict.items():
        region = Region(region_name, player, multiworld)
        set_region_locations(region, region_data.location_names, is_pool_expanded)
        multiworld.regions.append(region)
    connect_regions(multiworld, player)
```

Please do not use `world: MultiWorld`. It can get confusing, both for people reviewing the code, and people using it as a reference while developing their own apworlds.

**Review by ScipioWright — COMMENTED** [review]
Just a quick skim, looks pretty great so far. Will review it in more depth later.

File: worlds/saving_princess/Regions.py
Comment (LeonarthCG) [reply worlds/saving_princess/Regions.py:20]: When I first made the class there was more to it, but you are right that as it currently stands there's no reason for it to exist. I'll do away with it.

File: worlds/saving_princess/Rules.py
Comment (ScipioWright) [inline worlds/saving_princess/Rules.py:21]:
```suggestion
                        state.has_any({ITEM_WEAPON_FIRE, ITEM_WEAPON_ICE, ITEM_WEAPON_VOLT}, world.player)
```
You can use state.has_any to do basically what you're doing here

File: worlds/saving_princess/Rules.py
Comment (ScipioWright) [inline worlds/saving_princess/Rules.py:56]:
```suggestion
                    state.has_all({EP_ITEM_GUARD_GONE, EP_ITEM_CLIFF_GONE, EP_ITEM_ACE_GONE, EP_ITEM_SNAKE_GONE}, world.player)
```

You can use state.has_all to do the same thing you had here

File: worlds/saving_princess/Options.py
Comment (ScipioWright) [inline worlds/saving_princess/Options.py:119]:
```suggestion
@dataclass
class SavingPrincessOptions(PerGameCommonOptions):
    start_inventory_from_pool: StartInventoryPool
    expanded_pool: ExpandedPool
    trap_chance: TrapChance
    death_link: DeathLink
    instant_saving: InstantSaving
    sprint_availability: SprintAvailability
    cliff_weapon_upgrade: CliffWeaponUpgrade
    ace_weapon_upgrade: AceWeaponUpgrade
    shake_intensity: ScreenShakeIntensity
    iframes_duration: IFramesDuration
    music_shuffle: MusicShuffle
```

The order of the options in your dataclass determines what order the options show up in the player yamls and what order they show up in on the website. I recommend grouping the "generation" options together towards the top and the "client" options towards the bottom. The suggestion is my recommendation here.

**Review by ScipioWright — APPROVED** [review]
Small game, the apworld is written very cleanly. Unit tests passed fine, no errors when genning lots of random seeds. I did not test the client at all.

**Review by nicholassaylor — CHANGES_REQUESTED** [review]
# Nature of Review
* Only `Options.py` and the documentation was reviewed
* Documentation was checked through an editor and verified through WebHost

### Options
Looking over the options, I did not see anything of concern, the tooltips were self-explanatory and the types made sense.

### Documentation
Documentation, particularly in the game page is fairly good. I left some comments that I think will make the pages a bit more readable and condense information.

### Client
Looking over the set-up guide, I see that a `.bs4diff` file is used to patch the `data.win` file of a Gamemaker game. I see that there are concerns about assets and code being included, but you may want to look at the Undertale implementation for patching or AP Procedure Patch (APPP) to see if you can integrate the patching process into either a launcher for patching or a standalone program. I don't consider this a requirement for merging, but it may make the game more approachable for randomizing.

### Additional Housekeeping
The following things should also get done before merging
* Add game to `README.md`
* Add your implementation to `CODEOWNERS`

File: worlds/saving_princess/__init__.py
Comment (ScipioWright) [inline worlds/saving_princess/__init__.py:5]:
```suggestion
from worlds.AutoWorld import World, WebWorld
```

Imports from outside your world need to be absolute per the specification here: https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/apworld%20specification.md#caveats

File: worlds/saving_princess/Items.py
Comment (Exempt-Medic) [inline worlds/saving_princess/Items.py:25]: I don't believe this `if` block ever does anything because all your filler and victory items explicitly set their `count` to 0 and the default for `count_extra` is already 0

Comment (LeonarthCG) [reply worlds/saving_princess/Items.py:25]: I agree, it is redundant and I should decide on one or the other.

I personally think that keeping the `if` block (and adding the check for `ItemClass.trap` that it is missing) while removing the redundant arguments in the item data dicts would look cleaner.

File: worlds/saving_princess/docs/en_Saving Princess.md
Comment (LeonarthCG) [reply worlds/saving_princess/docs/en_Saving Princess.md]: I have read the different comments you gave me and tried to rewrite the info page so that it is more condensed and information is not duplicated.

File: README.md
Comment (nicholassaylor) [inline README.md:1]: This has a merge conflict and needs to be resolved

File: SavingPrincessClient.py
Comment (nicholassaylor) [inline SavingPrincessClient.py:1]: I believe that in general, core is trying to shift away from clients in the root AP directory and instead having them open through the launcher.

File: worlds/saving_princess/docs/setup_en.md
Comment (nicholassaylor) [inline worlds/saving_princess/docs/setup_en.md:16]: If you end up moving the client to launcher only, this would need to be changed

File: worlds/saving_princess/docs/setup_en.md
Comment (nicholassaylor) [inline worlds/saving_princess/docs/setup_en.md:65]:
```suggestion
You can customize your options by visiting the [Saving Princess Player Options Page](/games/Saving%20Princess/player-options).
```
Fix broken link

File: worlds/saving_princess/Client.py
Comment (nicholassaylor) [inline worlds/saving_princess/Client.py:189]:
```suggestion
    if not is_install_valid():
```
No need to hold onto this value since it's not used elsewhere

**Review by nicholassaylor — APPROVED** [review]
# Nature of Review
* All aspects of previous review were re-reviewed
* Client was briefly looked over

### Documentation
Changes look good, very minor changes suggested, otherwise very good

### Client
Client now has an automated patch feature while still retaining the option to patch manually. I only took a cursory look over the patching process and didn't see any glaring issues. I was not able to test patching as I do not own the game.

### Options
No changes since last review, however with #2614 being merged after my last review, it may be a good idea to take a look at Option Groups and Presets before merging. Not a requirement, but looking at the dataclass for your options, it seems like Option Groups would serve you well.

### Housekeeping
Game was added to `README.md` and `CODEOWNERS` but a merge conflict on `README.md` needs to be resolved. Minor inconvenience at most.

Overall, the concerns that I had previously were addressed and any remaining reservations are extremely minor and/or optional.

File: worlds/saving_princess/__init__.py
Comment (Exempt-Medic) [inline worlds/saving_princess/__init__.py:141]: You could use `self.options.as_dict("death_link", "expanded_pool" "instant_saving", "sprint_availability", "cliff_weapon_upgrade", "ace_weapon_upgrade", "shake_intensity", "iframes_duration")` instead here.

File: worlds/saving_princess/__init__.py
Comment (Exempt-Medic) [inline worlds/saving_princess/__init__.py:156]: This should probably be moved to line up with when it occurs in the call order, namely after `create_items` / before `fill_slot_data`

File: worlds/saving_princess/Rules.py
Comment (Exempt-Medic) [inline worlds/saving_princess/Rules.py:7]: Do you have to import `SavingPrincessWorld` then?

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
A lot of minor things and mostly docs. Overall this all looks very good, though some of the helper functions, like `get_location` confuse me. Only major thing was that you're putting events in the `name_to_id` lookups. I looked at everything except `Client.py` and have not done any generations yet

**Comment by Exempt-Medic** [comment]
I've now done one thousand generations (wow this game gens _fast_) and no errors occurred across 250 solo gens and 750 gens of increasing player size. Went over the WebHost display and everything looks fine there as well, you could potentially consider adding option groups or option presets, but it's small enough that it seems alright without.

**Comment by Exempt-Medic** [comment]
Having now played the game, it connects and seems to run just fine, however you're connecting with the `AP` tag which you likely shouldn't be doing. You also may want to add code in `generate_early` to check that the player name can actually be input into your game. Could do something like:
```py
from Options import OptionError, PerGameCommonOptions

    def generate_early(self) -> None:
        if not self.player_name.isascii():
            raise OptionError(f"{self.player_name}'s name must be only ASCII.")
```

File: worlds/saving_princess/Items.py
Comment (Exempt-Medic) [inline worlds/saving_princess/Items.py:33]: Instead of creating these every single time you create an item, why not just make lookup tables directly beforehand that are shared and then use those?

File: worlds/saving_princess/docs/en_Saving Princess.md
Comment (LeonarthCG) [reply worlds/saving_princess/docs/en_Saving Princess.md:41]: By cheated in I meant anything that has the server as the player that found it. So, yeah, start_inventory would count for that. I found it annoying to see all the pop-ups when using start_inventory, but perhaps this would just be confusing instead?

I don't feel strongly one way or the other, so if this is outside of the norm of what other worlds do then I'm happy to change it.

File: worlds/saving_princess/Options.py
Comment (Exempt-Medic) [reply worlds/saving_princess/Options.py:74]: Oh, having played the game now, this no longer seems weird XD

File: worlds/saving_princess/docs/setup_en.md
Comment (LeonarthCG) [reply worlds/saving_princess/docs/setup_en.md:105]: By "should sync up" I was trying to say that gm-apclientpp would take care of it.
I guess I should just say "will sync up"? Because if we assume the mod, the client and the protocol all work, then it will sync up, and if doesn't I don't think there's anything the user could really do to help it anyway?

The only way I've managed to make it not sync properly in my testing has been by wiping all of the local server's data and then starting it back up, but if it comes to that I'm not sure there's much a user can do about it.

File: worlds/saving_princess/Client.py
Comment (LeonarthCG) [reply worlds/saving_princess/Client.py:54]: No, I don't think it is.
I honestly cannot remember why I did that.

File: worlds/saving_princess/docs/setup_en.md
Comment (Exempt-Medic) [reply worlds/saving_princess/docs/setup_en.md:105]: Changing to "will" makes sense to me then

**Comment by LeonarthCG** [comment]
@Exempt-Medic 
> Something I forgot to comment on: The trap change is simply a flat chance, so if there are 10 filler items and 50% traps, there are not assuredly 5 traps, but there could be any number 0-10. Is this intended?

This is intentional, and I believe the docstring ("Likelihood of a filler item becoming a trap.") for the option is clear enough.

Is this design choice a big deal? I figured traps should be a surprise, and knowing the exact number of them kinda goes against that in my mind.

**Comment by Exempt-Medic** [comment]
> @Exempt-Medic
> 
> > Something I forgot to comment on: The trap change is simply a flat chance, so if there are 10 filler items and 50% traps, there are not assuredly 5 traps, but there could be any number 0-10. Is this intended?
> 
> This is intentional, and I believe the docstring ("Likelihood of a filler item becoming a trap.") for the option is clear enough.
> 
> Is this design choice a big deal? I figured traps should be a surprise, and knowing the exact number of them kinda goes against that in my mind.

I don't think it's a big deal, just a design choice that different devs do wildly differently so i figured it was worth asking about just in case you hadn't considered it.

File: worlds/saving_princess/docs/setup_en.md
Comment (LeonarthCG) [reply worlds/saving_princess/docs/setup_en.md:77]: This is what I came up with, I'm not sure if this is what you had in mind, so I'm asking before committing anything:

>After launching the game, enter the Archipelago options menu through the in-game button with the Archipelago icon.
>From here, enter the different menus and type in the following details in their respective fields:
>- **server:port** (e.g. `archipelago.gg:38281`)
>    * If hosting on the website, the server and port will be shown in your created room.
>- **slot name** (e.g. `Player`)
>    * This is your player name, which you chose along with your player options.
>- **password** (e.g. `123456`)
>   * If the room does not have a password, it should be left empty.
>
>This configuration persists through launches and even updates.

File: worlds/saving_princess/Rules.py
Comment (LeonarthCG) [reply worlds/saving_princess/Rules.py:57]: If expanded pool is off, the expanded pool items do nothing. The game will just ignore them in favor of checking for the flags for the corresponding event (mimicking the vanilla game).

So, (if I have this right) using `state.has_any` here would make it so the generation doesn't think the player needs to reach `EVENT_ITEM_POWER_ON` to get power back if they have `EP_ITEM_POWER_ON` in their starting inventory.

File: worlds/saving_princess/Rules.py
Comment (LeonarthCG) [reply worlds/saving_princess/Rules.py:7]: Not sure if you mean I'm missing the import (I think this is what other worlds do/what I'm missing):
```py
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  from . import SavingPrincessWorld
```
Or if you are asking why I'm using it in the first place?

File: worlds/saving_princess/Items.py
Comment (LeonarthCG) [reply worlds/saving_princess/Items.py:33]: Sorry, I'm not sure I follow.
I think you mean `count` and `count_extra`? But I'm pretty sure those are only touched once per entry in my item dictionary, which is the only place `ItemData` instances are made (unless I made a mistake somewhere).

Or maybe it's my class naming that is confusing? `ItemData` holds data used to create `SavingPrincessItem` instances, but the `SavingPrincessItem` class doesn't have `ItemData` in it.

File: worlds/saving_princess/Rules.py
Comment (Exempt-Medic) [reply worlds/saving_princess/Rules.py:7]: Yeah, I'm saying you'd want to include that exact code here

File: worlds/saving_princess/docs/setup_en.md
Comment (Exempt-Medic) [reply worlds/saving_princess/docs/setup_en.md:77]: That looks great to me!

File: worlds/saving_princess/Rules.py
Comment (Exempt-Medic) [reply worlds/saving_princess/Rules.py:57]: Ah they get completely ignored, this makes sense then

File: worlds/saving_princess/Items.py
Comment (Exempt-Medic) [reply worlds/saving_princess/Items.py:33]: I'm saying that every time you call `create_item` the `index` and `name` variables get totally rebuilt. But since the item dict is static, you could just create these once, right after making the item dict itself instead of creating them many times

Comment (LeonarthCG) [reply worlds/saving_princess/Items.py:33]: I understand now. Maybe something like this would make more sense:
```py
    def create_item(self, player: int):
        return SavingPrincessItem(item_data_names[self], self.item_class, self.code, player)

[... all of the ItemData dict stuff ...]

item_data_names: Dict[ItemData, str] = {value: key for key, value in item_dict.items()}
````

Comment (Exempt-Medic) [reply worlds/saving_princess/Items.py:33]: Looks good to me

**Comment by LeonarthCG** [comment]
Unless I missed something, I believe I tackled all of the concerns.

I also removed the standalone launcher, leaving only the ArchipelagoLauncher one, and added option groups and presets.

**Review by Exempt-Medic — APPROVED** [review]
Recent changes LGTM, did 100 more singleplayer generations and 100 more 25-player generations. Tested playing through one game. All prior comments were addressed.

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
Incredibly nice implementation. I have one important request and two minor comments that aren't important.

File: worlds/saving_princess/__init__.py
Comment (NewSoupVi) [reply worlds/saving_princess/__init__.py:20]: Fellow Core Maintainer Black Sliver agrees with me that this is a good procedure for now.

Comment (LeonarthCG) [reply worlds/saving_princess/__init__.py:20]: I think I should be able to implement passing the link to the game as well, so I'll look into that.

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**
